cache/lru_cache.cc

```diff
@@ -241,17 +258,42 @@ void LRUCacheShard::LRU_Insert(LRUHandle* e) {
     e->prev->next = e;
     e->next->prev = e;
     e->SetInHighPriPool(true);
+    e->SetInLowPriPool(false);
     high_pri_pool_usage_ += e->total_charge;
     MaintainPoolSize();
-  } else {
+  } else if (low_pri_pool_ratio_ > 0 && (e->IsLowPri() || e->HasHit())) {
     // Insert "e" to the head of low-pri pool. Note that when
     // high_pri_pool_ratio is 0, head of low-pri pool is also head of LRU list.
-    e->next = lru_low_pri_->next;
-    e->prev = lru_low_pri_;
+    if (lru_low_pri_ == &lru_) {
+      // The low-pri pool is empty and never have an entry inserted into it yet.
+      e->next = lru_bottom_pri_->next;
+      e->prev = lru_bottom_pri_;
+    } else {
+      // The low-pri pool is not empty.
+      e->next = lru_low_pri_->next;
+      e->prev = lru_low_pri_;
+    }
```

Comment:
Still not sure about this... The way I read the code, `lru_low_pri_ == &lru_` means that there are no items in the *high-priority* pool.

Reply:
I'm wondering if we could avoid having to branch here (and maintain the invariant that `lru_low_pri_` unconditionally denotes the position where low-priority items are to be inserted) by changing the condition on the bottom-priority (`else`) branch on line 294? 

Side note: notions of head and tail are somewhat confusing in the LRU cache code. For instance, on mainline we have this:

```
    // Insert "e" to the head of low-pri pool. Note that when
    // high_pri_pool_ratio is 0, head of low-pri pool is also head of LRU list.
    e->next = lru_low_pri_->next;
    e->prev = lru_low_pri_;
```

I would personally say `lru_low_pri_` is the tail of the low-priority pool, not the head, for two reasons: 1) the old `lru_low_pri_` becomes the `prev` of the new element, and 2) `lru_low_pri_` denotes the low-priority element that gets evicted last (not first).

Anyways, maybe it's just me, and it's OK to leave the terminology as-is for now.

Reply:
Thanks!!!